src/test/ui/rfc-2632-const-trait-impl/const-trait-bound-opt-out/in-impl-trait.rs

```diff
@@ -0,0 +1,21 @@
+#![feature(const_trait_bound_opt_out)]
+#![feature(associated_type_bounds)]
+#![allow(incomplete_features)]
+
+trait T {}
+struct S;
+impl T for S {}
+
+fn rpit() -> impl ?const T { S }
```

Comment:
Seems like we came away with different interpretations of what Oliver said.
I think this would be perfectly well defined and semantically sane (same with `dyn ?const Trait`).

After some more reflection, I think the semantic property one would want is that `?const` is legal if and only if we are in a `const`-by-default (`const fn`) or `const Trait for { ... }`. However, if you want to get this landed quicker, I'm fine with this for now (but leave the comment visible).

Reply:
@oli-obk ^

Reply:
While these would be fine in the future, having them be a hard error for now and giving them their own feature gates in the future seems the best way forward to me. The RFC specifically talked about a minimal version that does not include `impl Trait` at all

Reply:
After porting these checks to AST validation, I also decided to error when they are found on trait bounds (e.g., `trait T: ?const Super {}`). While this is not explicitly called out as a future extension by the RFC, it will not be useful until `const` trait definitions have some meaning.